plugins/cairo-lang-macro/src/lib.rs

```diff
@@ -53,22 +57,70 @@ impl Display for AuxData {
     }
 }
 
+/// Diagnostic returned by the procedural macro.
+#[derive(Debug)]
+pub struct Diagnostic {
+    pub message: String,
+    pub severity: Severity,
+}
+
+/// The severity of a diagnostic.
+#[derive(Debug)]
+pub enum Severity {
+    Error,
+    Warning,
+}
+
+impl Diagnostic {
+    pub fn error(message: impl ToString) -> Self {
+        Self {
+            message: message.to_string(),
+            severity: Severity::Error,
+        }
+    }
+
+    pub fn warn(message: impl ToString) -> Self {
+        Self {
+            message: message.to_string(),
+            severity: Severity::Warning,
+        }
+    }
+}
+
 impl ProcMacroResult {
     /// Convert to FFI-safe representation.
     ///
     /// # Safety
     #[doc(hidden)]
     pub fn into_stable(self) -> StableProcMacroResult {
         match self {
-            ProcMacroResult::Leave => StableProcMacroResult::Leave,
-            ProcMacroResult::Remove => StableProcMacroResult::Remove,
+            ProcMacroResult::Leave { diagnostics } => {
+                let (ptr, n) = unsafe { Diagnostic::allocate(diagnostics) };
```

Comment:
could you please elaborate why do you need these allocate/deallocate functions, when you can pass Box etc. via FFI boundary, which have Drop implemented etc?

Reply:
Sure, apparently while Box is safe to pass, a slice is not. There even is a clippy lint.
```
warning: `extern` fn uses type `[cairo_lang_macro_stable::StableDiagnostic]`, which is not FFI-safe
  --> scarb/src/compiler/plugin/proc_macro/ffi.rs:75:19
   |
75 | type ExpandCode = extern "C" fn(StableTokenStream) -> StableProcMacroResult;
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not FFI-safe
   |
   = help: consider using a raw pointer instead
   = note: slices have no C equivalent
   = note: `#[warn(improper_ctypes_definitions)]` on by default
```
I've also tried `Box<[Box<str>]>` with the same results 😢 
So next best thing for me is to use `into_raw` etc. 
What do you think? 🤔

Reply:
Maybe try the following?
1. Wrap the pointers into a struct to isolate this unsafety. Like this: https://docs.rs/ffi-std/latest/src/ffi_std/slice.rs.html (though don't pull this as a library)
2. If needed, use [`ManuallyDrop`](https://doc.rust-lang.org/stable/std/mem/struct.ManuallyDrop.html) to pass these things via FFI and still control deallocation via `Drop`.

Reply:
Great, I'll try right away!